traits/trait_types.py

```diff
@@ -169,20 +170,6 @@ def _validate_int(value):
         return int(operator.index(value))
 
 
-def _validate_float(value):
-    """ Convert an arbitrary Python object to a float, or raise TypeError.
-    """
-    if type(value) is float:  # fast path for common case
-        return value
-    try:
```

Comment:
This is where the bug is: this code misses the chance to convert objects that have an `__index__` method.